src/interfaces/python/swig_typemaps.i

```diff
@@ -1373,8 +1381,40 @@ def _internal_factory_wrapper(object_name, new_name, docstring=None):
     return _internal_factory
 
 for factory in _FACTORIES:
-	_rename_python_function(sys.modules[__name__], factory[1], factory[0])
-	_swig_monkey_patch(sys.modules[__name__], factory[1], _internal_factory_wrapper(*factory))
+	# renames function in the current module (shogun) from `factory` to "_" + `factory`
+	# which "hides" it from the user
+	factory_private_name = "_{}".format(factory)
+	_rename_python_function(sys.modules[__name__], factory, factory_private_name)
+	# adds a new function called `factory` to the shogun module which is a wrapper
+	# that passes **kwargs to objects via .put (see _internal_factory_wrapper)
+	_swig_monkey_patch(sys.modules[__name__], factory, _internal_factory_wrapper(factory_private_name, factory))
+
+for getter in _GETTERS:
+	_rename_python_function(_shogun.SGObject, getter, "_{}".format(getter))
+
+def _internal_get_param(self, name):
+	"""
+    Returns the value of the given parameter.
+    The return type depends on the parameter,
+    e.g. could be a builtin scalar or a
+    numpy array representing a vector or matrix
+   	"""
+
+	for f in (self._get,
+			  self._get_real,
+			  self._get_int,
+			  self._get_real_matrix,
+			  self._get_real_vector,
+			  self._get_int_vector):
+		try:
+			return f(name)
+		except SystemError:
+			pass
+		except Exception:
+			raise
+	raise KeyError("There is no parameter called '{}' in {}".format(name, self.get_name()))
```

Comment:
There is an issue here, which @avramidis ran into: If none of the getters above lead to a result, then the method will claim that the parameter doesnt exist. But in his case, this is since there was not internal `get_char_string_list` method defined yet, thus none of the methods here matched. What happened then is that `obj.parameter_names()` showed a name, but `obj.get(name)` claimed that the parameter didn't exist. I wonder what is the best (most user/dev friendly) solution for this: Write down the types that were tried? Bad since those could be many and this confuses the user. The current way? Confuses people if a getter is missing. I would be afraid that this pops up a few years in the future (someone adds a new type) and then someone has to spend hours figuring this out. Any suggestions?

Reply:
In that case can add a check `name in obj.parameter_names()`, and if it is say that there is no getter in the Python API that can be used for it?

Reply:
But only if we reach the end of the checks

Reply:
> In that case can add a check `name in obj.parameter_names()`, and if it is say that there is no getter in the Python API that can be used for it?

yes that would be a minimal solution at least. Then a dev will immediately be able to figure this out

Reply:
But actually, being able to ask what type a parameter is (in string form) might also be something that could be useful or?

Reply:
Yes, but how would it work? With the C++ demangling?

Reply:
Is it possible with that? I’m not sure. But at least what you wrote above is good: two messages, either saying Param doesn’t exist (and actually then no need to try to dispatch) or say that python api doesn’t support getting it.

This is in fact independent of whether we can get a type string, which would just make it easier to track the source

Reply:
OK, I added the check to raise an error if the param exists but can't be accessed with any of the known getters. 
It would be possible to get the type string I think, just need to have a method that does it in CSGObject (if there isn't one already?)

Reply:
Yes let's put that in. I think it should be part of the swig code, not the libshogun (not useful there)